veg2hab/functionele_samenhang.py

```diff
@@ -0,0 +1,295 @@
+    )
+
+    return clusters
+
+
+def _extract_elmid_perc_habtype(gdf: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+    """
+    Trekt uit de vegkartering gdf de benodigde informatie voor het bepalen van clusters
+    Ieder complexdeel/HabitatKeuze krijgt een eigen rij in de output, met de identifier (ElmID, complex-deel-index)
+    Als er meerdere complexdelen met hetzelfde habtype in een vlak zitten, worden deze samengevoegd
+    """
+
+    def apply_func(row: pd.Series) -> pd.DataFrame:
+        # identifier (Elmid, [cmplxdeel_n]) | percentage | habitattype | geometry
+        identifier = []
+        percentage = []
+        habtype = []
+        geometry = []
+        for idx, keuze in enumerate(row["HabitatKeuze"]):
+            # Dit stelt ons in staat weer terug te gaan naar de originele habitatkeuze
+            identifier.append((row["ElmID"], (idx,)))
+            # Nodig voor het bepalen van de buffergrootte
+            percentage.append(row.VegTypeInfo[idx].percentage)
+            # We clusteren binnen ieder habtype
+            habtype.append(keuze.habtype)
+            # We kunnen niet clusteren zonder geometrie
+            geometry.append(row.geometry)
+
+        gdf = gpd.GeoDataFrame(
+            {
+                "identifier": identifier,
+                "percentage": percentage,
+                "habtype": habtype,
+                "geometry": geometry,
+            }
+        )
+
+        # Checken op meerdere complexdelen met hetzelfde habtype binnen het huidige vlak
+        if len(set(habtype)) != len(habtype):
+            non_unique_habtypes = set(
+                habtype[i] for i in range(len(habtype)) if habtype.count(habtype[i]) > 1
+            )
+            # We vervangen alle rijen met hetzelfde habtype door 1 rij met gecombineerde identifier
+            # en de som van de percentages
+            for type in non_unique_habtypes:
+                all_rows_of_type = gdf[gdf.habtype == type].copy()
+                identifiers = all_rows_of_type["identifier"].tolist()
+                new_row = gpd.GeoDataFrame(
+                    {
+                        "identifier": [
+                            (
+                                identifiers[0][0],
+                                tuple(identifier[1][0] for identifier in identifiers),
+                            )
+                        ],
+                        "percentage": [sum(all_rows_of_type.percentage)],
+                        "habtype": [type],
+                        "geometry": [all_rows_of_type.geometry.iloc[0]],
+                    }
+                )
+                gdf = gdf[gdf.habtype != type]
+                gdf = pd.concat([gdf, new_row], ignore_index=True)
+
+        return gdf
+
+    extracted = gdf.apply(apply_func, axis=1)
+    return pd.concat(extracted.to_list(), ignore_index=True)
+
+
+def _remove_habtypen_due_to_minimum_oppervlak(
+    gdf: gpd.GeoDataFrame, to_be_edited: Set[Tuple[int, Tuple]]
```

Comment:
is deze to_be_edited, niet gewoon een mapping? oftewel een dict?

Reply:
Je bedoelt dat het ElmID naar complex-deel indices omzet? Je kunt meerdere keren hetzelfde ElmID hebben met steeds een ander complexdeel ((ElmID 123, idx 0), (ElmID 123, idx 1)), en je kunt meerdere ElmID's hebben met hetzelfde complexdeel ((ElmID 123, idx 0), (Elmid 234, idx 0)). Dan kun je er geen dict van maken.

Reply:
Het is nu toch to_be_edited: Set[FuncSamenhangID]